Gametest verification

supersaiyansubtlety opened the pull request:
Adds a single gametest using the local mappings produced by the mappings project.

Adds a step to the Build workflow that runs `build runGametest` in the gametest directory.

Meant to catch things like #695 and #696.
Building also catches things like #698.

The gametest:
- (implicitly) verifies bootstrap bootstraps
- (implicitly) loads a world
- spawns and kills each entity type

~~I could add a non-game test to explicitly verify bootstrap bootstraps so it fails earlier, but I'm not sure if it's worth it.~~
Can't do that without fapi.

~~This moves basically the entire mappings project into an includeBuild'ed `core` project.~~

~~As a result, all the mappings project tasks must be prefixed with `:core:`, and actions will need to be updated.~~

Thanks to @lukebemish for help with getting loom to use local mappings!

supersaiyansubtlety committed:
move prior root project to :core subproject

supersaiyansubtlety committed:
attempt to setup gametest subproject

supersaiyansubtlety committed:
make :core and included build instead of a subproject and move :buildSrc into :core
can't get :gametest to see the mappings :core creates

supersaiyansubtlety committed:
use :core qsl dependency substitution

supersaiyansubtlety committed:
try publishing component

supersaiyansubtlety committed:
rename gametestVerify publication

supersaiyansubtlety committed:
try publishing to local dir and dummy plugin

supersaiyansubtlety committed:
specify intermediary-v2 classifier

supersaiyansubtlety committed:
it configures!

supersaiyansubtlety committed:
rename build stuff gametestVerify -> gametestable
add predicates so publish task only publishes gametestable to local repo
cleanup

supersaiyansubtlety committed:
rename dummy plugin -> gametestable
expose :core project version to :gametest project via GametestableExtension

supersaiyansubtlety committed:
switch to floom because qloom only provides 1.8 and fapi requires 1.10
update gradle to 8.12
setup gametest mod with TestEntitiesSpawn

lukebemish commented:
Is there a reason the root project was moved to a subdirectory? Why not keep it where it is and have the game tests in a subdirectory?

supersaiyansubtlety commented:
> Is there a reason the root project was moved to a subdirectory? Why not keep it where it is and have the game tests in a subdirectory?

Since the gametest project includeBuild’s the core project, I’m not sure how I could keep the core project in the root directory without creating circular project dependencies. 
I’m also not sure how I’d refer to it for includeBuild, does it accept relative paths like ‘..’?

lukebemish commented:
Core project in root project, with no dependency on the gametest project at all. Gametest project includebuilds it's parent directory: `includeBuild '../'`.

supersaiyansubtlety commented:
If the gametest project is neither `include`d nor `includeBuild`ed, how is it recognized by gradle at all?
Or are you saying we change the working directory to the gametest project before running the gametest task?

lukebemish commented:
Just run the task from that directory, yes. GH actions have a working directory option -- you can reference `../gradlew` just fine within there. There's really no good reason for that test to be part of the main build given how this all works.

supersaiyansubtlety committed:
move core back to the root project
rename dummyPlugin -> gametestable

supersaiyansubtlety committed:
add gradle-wrapper.jar files

supersaiyansubtlety committed:
restore files with whitespace diffs

lukebemish commented on build.gradle:
```diff
@@ -324,6 +367,13 @@ publishing {
 		}
+
+		// for gametest project
+		gametestable(MavenPublication) {
+			groupId 'org.quiltmc'
+			artifactId 'gametestable-quilt-mappings'
+			from components.gametestable
+		}
```
Is there a reason the gametest project can't just consume the _normal_ publication published to that local maven? After all, it's all running in an earlier build step so you don't actually need the component/consumable variant stuff.

supersaiyansubtlety replied:
Oh was the component only required for dependency substitution?
I'll try without it.

supersaiyansubtlety replied:
Unless I made a mistake, it seems it needs the component publishing for gametest to resolve it
<https://github.com/supersaiyansubtlety/quilt-mappings/tree/gametest-verification-no-component-publishing>

>FAILURE: Build failed with an exception.
> 
> * What went wrong:
> A problem occurred configuring root project 'gametest'.
> > Could not resolve all dependencies for configuration ':mappings'.
   > Could not find org.quiltmc:gametestable-quilt-mappings:1.21.5+build.local.
     Searched in the following locations:
       ...
       - file:/C:/Users/super/Projects/Minecraft/Modding/Quilt/quilt-mappings/build/gametestable/org/quiltmc/gametestable-quilt-mappings/1.21.5+build.local/gametestable-quilt-mappings-1.21.5+build.local.pom
     Required by:
         root project :

supersaiyansubtlety replied:
oh duh, I forgot to remove `gametestable-` from the dependency

supersaiyansubtlety committed:
publish normal maven publication to local gametestable repo instead of separate component-based publication; gametest can't resolve it

supersaiyansubtlety committed:
remove 'gametestable-' prefix from gametest mappings dep, cleanup

lukebemish commented on gametest/build.gradle:
```diff
@@ -0,0 +1,52 @@
+    filesMatching('fabric.mod.json') {
+        expand expanding
+    }
+}
```
Trailing newline please? I know, minor thing, probably no big deal...

lukebemish reviewed (approved):
Gradle-wise, looks good to me!

supersaiyansubtlety committed:
add a couple trailing newlines

github-actions[bot] committed:
Merge branch '25w15a' into gametest-verification

eristhea commented:
just wanted to pop in and say this is INCREDIBLE i'm so amazed that you cooked this up

i'm still a bit too busy to really dig into reviewing, but i appreciate this work sm and i'll give it a look within a week or so!

github-actions[bot] committed:
Merge branch '25w16a' into gametest-verification

eristhea committed:
update to latest snapshot

eristhea reviewed (changes requested):
i would prefer that these versions were situated in the same file as all the other versions. that way we can more easily auto-update them in matcher!

also i had to remove the mc version constraint since it didn't work on snapshots (floader messes with the version number, making the snapshot `1.21.5-alpha.25.w.16.a` or something like that). i don't see a reason to fix the constraint so i just removed it

supersaiyansubtlety committed:
move gametest's libraries to root's libs.version.toml

supersaiyansubtlety commented:
thanks for catching the snapshot issue!

supersaiyansubtlety committed:
update build workflow to use runGametest instead of javadocJar (which will still run since it's a task dependency)

supersaiyansubtlety committed:
make gametest gradle executable

supersaiyansubtlety committed:
restore root's build javadocJar as a seprate, explicit step

supersaiyansubtlety commented:
got permissions sorted, so I've updated the build workflow with a step that runs `build runGametest` in the gametest directory

supersaiyansubtlety committed:
restore --stacktrace to root build command

github-actions[bot] committed:
Merge branch '25w17a' into gametest-verification

actions-user committed:
Drop invalid mappings

eristhea reviewed (approved):
now that i'm thinking of it, do we maybe want to query the FAPI version at runtime and cancel if no version exists?
this fixes the issue of FAPI not always being available to update during matching (which would mean we have to go in later for every snapshot and update the version), but it might introduce a lack of reproducibility

supersaiyansubtlety commented:
~~Hm, I wonder how doable it would be to have our own mixin-based entry point and use vanilla's data-driven stuff
That would avoid the issue entirely (and I could switch back to qloader).~~

~~I'll look into that soon.~~

Done!

supersaiyansubtlety committed:
use mixin and data instead of fapi for gametest

supersaiyansubtlety committed:
remove fapi entirely
add MainMixin to bootstrap gametest functions
add task to copy data to world datapack because test data isn't loaded from mod resources without fapi

supersaiyansubtlety committed:
convert back to quilt mod \[T]/

supersaiyansubtlety commented on gametest/build.gradle:
```diff
@@ -0,0 +1,75 @@
+    }
+}
+
+// mod test data isn't loaded without fapi, so copy it to a datapack
```
Just pointing this out to make sure we're ok with this additional jank.

This is easier and hopefully more maintainable than a mixin to load it from resources.

supersaiyansubtlety committed:
move gametest datapck out of resources

github-actions[bot] committed:
Merge branch '25w18a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w19a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w20a' into gametest-verification

supersaiyansubtlety commented:
Something in one of the recent snapshots changed things so the gametest server hangs, I'm looking into it.

supersaiyansubtlety committed:
make test server tick

supersaiyansubtlety commented:
> Something in one of the recent snapshots changed things so the gametest server hangs, I'm looking into it.

fixed

supersaiyansubtlety commented:
Note that currently, due to the way loom caches things, it's often necessary to use `--refresh-dependencies` when building or running the gametest locally. If you've previously built, then update mappings, those changes won't be reflected without `--refresh-dependencies`.

This isn't a problem for the build workflow because of containerization.

lukebemish commented:
I tracked down why that happens -- it's because the gametest build isn't pulling a new version of the relevant stuff from the dummy maven it's published to. That should be fixable by marking the dependency as changing -- literally just `changing = true` on the dependency.

supersaiyansubtlety committed:
prevent caching of gametest mappings dependency

supersaiyansubtlety commented on gametest/build.gradle:
```diff
@@ -0,0 +1,68 @@
+    }
+}
+
+configurations.mappings {
+    resolutionStrategy {
+        // do not cache mappings
+        cacheChangingModulesFor 0, SECONDS
+    }
+}
+
+dependencies {
+    minecraft(rootLibs.minecraft)
+    mappings ("org.quiltmc:quilt-mappings:${gametestable.version}:intermediary-v2") {
+        changing = true
+    }
```
This unfortunately didn't fix it.

github-actions[bot] committed:
Merge branch '25w21a' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.6-pre1' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.6-pre2' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.6-pre3' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.6-pre4' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.6-rc1' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.6' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.7-rc1' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.7-rc2' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.7' into gametest-verification

OroArmor reviewed (approved):
Looks good. Do we need the second copy of gradle?

supersaiyansubtlety committed:
update gametest code for new mappings

supersaiyansubtlety committed:
remove gametest's separate gradle instance and make it use root's instead

supersaiyansubtlety commented:
> Do we need the second copy of gradle?

Nope, removed

github-actions[bot] committed:
Merge branch '1.21.8-rc1' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.8' into gametest-verification

supersaiyansubtlety committed:
update for new mapping

github-actions[bot] committed:
Merge branch '25w31a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w32a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w33a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w34a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w34b' into gametest-verification

github-actions[bot] committed:
Merge branch '25w35a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w36a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w36b' into gametest-verification

github-actions[bot] committed:
Merge branch '25w37a' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.9-pre1' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.9-pre2' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.9-pre3' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.9-pre4' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.9-rc1' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.9' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.10-rc1' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.10' into gametest-verification

github-actions[bot] committed:
Merge branch '25w41a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w42a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w43a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w44a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w45a' into gametest-verification

github-actions[bot] committed:
Merge branch '25w46a' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.11-pre1' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.11-pre2' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.11-pre3' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.11-pre4' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.11-pre5' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.11-rc1' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.11-rc2' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.11-rc3' into gametest-verification

github-actions[bot] committed:
Merge branch '1.21.11' into gametest-verification

supersaiyansubtlety committed:
update for mapping and version changes
